README.md
# Rubicon Reminders

### A simple note-taking app.

Think 'sticky notes' but slightly more organized.

A place for all of your quick notes, reminders, or lists. All viewable in a single, unobtrusive interface and stored locally.

<img src="https://i.gyazo.com/c0713b447a39f10e9ff8cada30aa1533.png" width="400">

Made using tkinter.

## Installation 

* Either download the pre-packaged exe [here](https://github.com/ZG34/RubiconReminders/releases),
* Or package yourself using pyinstaller main.py -w , [source code on github](https://github.com/ZG34/RubiconReminders),
* Or run main.py file.
* May also run the CLI by running the rubicon_reminders_cli file. Note that these programs do not share notes between each other.


CHANGELOG.MD
# Changelog
### v0.3 - 1/28/22
* Added - TODO functionality, accessible via the note interface button 'Open TODO.'
* Added - hotkey functions for most major actions in both note and todo interfaces.
* Updated - Help menu


### v0.21 - 1/22/22
* Added - text wrapping on word in text area.
* Minor wording change on help menu.


### 1/21/22
* Added - text wrap on word in text output area.

### v0.2 - 1/20/22
* Added - [View Note Folder] menu item
* Added - [Help] menu item, which display keybinds and app functionality.
* Added - GUI display to show user that text file is being saved properly.


* Bugfix - Can now click anywhere and still manage to save to correct file. 
  * Previously, double-clicking away from listbox selection would break ability to save.

* Changed - changed program name.

###1/17/22
* Added - hotkey [ctrl-s] to save file text


* Bugfix - now able to properly delete text and save. 
  * Previously, deleted text would not register as deleted when saving.

### 1/14/22
* Added - scrollbar for text area
* Refactored code with more descriptive variable and function names

### v0.1 - 1/13/22
* First iteration of note taking GUI (main)
  * Create note folder
  * Create or delete new note(txt) files
  * Display contents of note folder in listbox
  * View contents of note in text area
  * Save additions to text area to opened note file

### v0.0 - 1/4/22
* Added functional command line script which: 
    * Prompts user for input to open existing local file or create new local note file.
    * Takes note entry, creating a new line of text per entry and finishing on 'done' entry.
    * Dates and closes note on 'done' entry.
    * On existing note open, allows 'done' entry and prints to file 'reviewed' with date time stamp.
* Project start. 


todo_tab.py
# this file holds the functionality for the to-do interface, inside a notebook framework which may be scaled to include
# the main note-taking app in another notebook tab.

from tkinter import (
    Tk,
    Frame,
    Text,
    Button,
    Checkbutton,
    Entry,
    Label,
    Listbox,
    END,
    Menu,
)
from tkinter.ttk import Notebook
import json
from datetime import datetime

# button labels as constants
ADD_NEW = "^ Enter & Add TODO ^"
ARCHIVE = "^ Archive Entry v"
DEL_ENTRY = "-Remove Archive-"

# reference file to save and restore state of listbox between instances, using JSON.
FILE1 = "todo_active"
FILE2 = "todo_archive"


class NotebookMain(Tk):
    def __init__(self, parent, *args, **kwargs):
        self.parent = parent
        self.window = Frame(self.parent)

        def notebook_assets(self):
            self.notebook1 = Notebook(
                self.parent)
            self.page1 = Frame(self.notebook1, bg="#000000")
            #self.page2 = Frame(self.notebook1)
            self.notebook1.add(self.page1, text="TODO")
            #self.notebook1.add(self.page2, text="NOTES")
            self.notebook1.pack()

            def todo_screen():
                def column1():
                    def add_entry(event):
                        todo = entry_1.get()
                        active_lbox.insert(
                            "end", todo)

                    def archive_entry(event):
                        selection = active_lbox.curselection()[0]
                        grabber = active_lbox.get(selection)
                        archive_lbox.insert("end", grabber)
                        active_lbox.delete(selection)

                    def de_archive_entry(event):
                        selection = archive_lbox.curselection()[0]
                        grabber = archive_lbox.get(selection)
                        active_lbox.insert("end", grabber)
                        archive_lbox.delete(selection)

                    def delete_active(event):
                        selection = active_lbox.curselection()[0]
                        active_lbox.delete(selection)

                    def delete_archive(event):
                        selection = archive_lbox.curselection()[0]
                        archive_lbox.delete(selection)

                    def save_state(event):
                        note = active_lbox.get(0, "end")
                        note2 = archive_lbox.get(0, "end")
                        with open(FILE1, "w") as f:
                            json.dump(note, f)
                            # json.dump(note2, f)
                        with open(FILE2, "w") as f2:
                            json.dump(note2, f2)
                        dt = datetime.now()
                        timestamp = f"{dt.hour}:{dt.minute}"
                        save_confirm.delete("1.0", "end")
                        save_confirm.insert("1.0", "TODO saved: " + str(timestamp))

                    def load_state():
                        with open(FILE1, "r+") as f:
                            data = json.load(f)
                            for v in data:
                                active_lbox.insert(END, v)
                        with open(FILE2, "r+") as f2:
                            data = json.load(f2)
                            for v in data:
                                archive_lbox.insert(END, v)

                    entry_1 = Entry(self.page1, highlightbackground="#9E9E9E", highlightthickness=1)
                    entry_1.grid(row=2, column=1, ipadx=20, pady=3)
                    entry_1.bind("<Tab>", add_entry)

                    active_lbox = Listbox(self.page1, width=18, highlightbackground="#9E9E9E",
                                          highlightthickness=1, font=("RobotoRoman", 12))
                    active_lbox.grid(row=4, column=1, pady=3)
                    active_lbox.bind("<Tab>", archive_entry)
                    active_lbox.bind("<Control-s>", save_state)
                    active_lbox.bind("<Delete>", delete_active)



                    archive_lbox = Listbox(self.page1, height=3, width=27, highlightbackground="#9E9E9E",
                                           highlightthickness=1, fg="#606060")
                    archive_lbox.grid(row=6, column=1, pady=3)
                    archive_lbox.bind("<Tab>", de_archive_entry)
                    archive_lbox.bind("<Delete>", delete_archive)
                    archive_lbox.bind("<Control-s>", save_state)

                    add_entry_button = Button(self.page1, text=ADD_NEW, command=lambda: add_entry(entry_1), width=20,
                                              relief="raised")
                    add_entry_button.grid(row=3, column=1, pady=3)

                    archive_entry_button = Button(self.page1, text=ARCHIVE, command=lambda: archive_entry(active_lbox),
                                                  width=20, relief="raised")
                    archive_entry_button.grid(row=5, column=1, pady=3)

                    delete_archived_button = Button(
                        self.page1, text=DEL_ENTRY, command=lambda: delete_archive(archive_lbox),
                        width=20, relief="raised")
                    delete_archived_button.grid(row=7, column=1, pady=3)

                    save_button = Button(self.page1, text="SAVE TODO", command=lambda: save_state(self.window))
                    save_button.grid(row=9, column=1)
                    save_button.config(font=("RobotoRoman", 11, "bold"), width=20, bg="#8E001C", fg="#FFB302")

                    save_confirm = Text(self.page1, height=1, width=23)
                    save_confirm.grid(row=8, column=1, pady=2)

                    try:
                        load_state()
                    except FileNotFoundError:
                        pass

                column1()

            todo_screen()

            def note_screen():
                pass

        notebook_assets(self)


def run():
    root = Tk()
    app = NotebookMain(root)
    root.mainloop()

# running here and on main.py resulted in break
# if __name__ == "__main__":
#     run()

menu_contents.py
from tkinter import Tk, Label
from os import startfile, path, getcwd, chdir


def help_screen():
    window = Tk()
    window.geometry("250x422")
    window.title("Help")

    bindings = Label(window, text="---Note Bindings--- \n \n"
                                  "[Ctrl-s] = Save Note \n"
                                  "[Del] = Delete selected note file \n"
                                  "[Enter] = Add New Note w/ Entered Title \n \n "
                                  "---TODO Bindings--- \n \n"
                                  "[Ctrl-s] = Save Lists \n"
                                  "[Del] = Delete selected archived todo \n"
                                  "[Tab] = Add new todo \n"
                                  "[Tab] = Move todo between active/archive ")
    bindings.pack(pady=10)

    functionality = Label(window, text="\n ---USER TIPS--- \n \n"
                                       " Add an entry by entering it above \n "
                                       "the 'add new' button,\n then click button or hotkey. \n \n"
                                       " Note Edits must be saved \n [Save Button, or Ctrl-S] \n before clicking a "
                                       "different "
                                       "note from the list.\n Otherwise, text will be lost. \n \n"
                                       "")
    functionality.pack(pady=10)

    window.mainloop()


def view_note_folder():
    parent_path = path.dirname(getcwd())
    chdir(parent_path)
    startfile('note_folder')
    # return to note directory in application after opening to view in explorer
    chdir('note_folder')



strat_notes_CLI.py
# to RUN APP IN CMD PROMPT: cd to StratNotes directory, or place strat_notes in default CMD directory:
# then run 'python strat_notes_terminalvers.py'

from os import listdir
from datetime import datetime

# this assigns dt variable as date + timestamp
dt = (datetime.now())
# establishing 0 as start for the numeration of each noted item per session
# n = 1
# n += 1            # TODO numerate note items per entry


# open existing or create a new file prompt:
p1 = input("(V)iew or (N)ew [V/N]: ").upper()
if p1 == "V":
    # this views file directory of existing notes if first input is (view)
    for file in listdir():
        if file.endswith(".txt"):
            print(file)
    # below opens existing file, allows multiple note lines, and dates it when finished with session.
    old_file = (input("which file would you like to open: "))
    hdl = open(old_file + ".txt", "r+")  # using r+ by default places text at beginning, overwriting.
    for line in hdl:  # as long as you first READ the file, then r+ becomes APPEND TO END.
        print(line.strip())
    of_note = input("Add Note: ")
    if of_note == "done":  # FIXME add accept on any 'done' check for upper and lowercase
        # specifies notes were reviewed if first note entry is "done"
        hdl.write(" REVIEWED: ")
        hdl.write(str(dt))
    # if first entered note is not 'done', continue asking for more notes until entry == 'done'
    else:
        hdl.write('\n')
        hdl.write(of_note)
        hdl.write('\n')
        while of_note != "done":
            of_note = input("Add more notes: ")
            while of_note != "done":
                hdl.write(of_note)
                hdl.write('\n')
        else:
            hdl.write("SESSION END: ")
            hdl.write(str(dt))
            hdl.write('\n')
        hdl.close()

# below is the block for generating and noting in a new file, if line 15 == 'N'
elif p1 == "N":
    new_file = input("new file name: ")
    hdl = open(new_file, "a")
    nf_note = input("Add Note: ")
    if nf_note == "done":
        print("finished")
    else:
        hdl.write(nf_note)
        hdl.write('\n')
        while nf_note != "done":
            nf_note = input("Add more notes: ")
            while nf_note != "done":
                hdl.write(nf_note)
                hdl.write('\n')
                break
        else:
            hdl.write("SESSION END: ")
            hdl.write(str(dt))
            hdl.write('\n')
    hdl.close()
else:
    print("Error: please enter V or N")



rubicon_reminders_cli.py
# the comments in this file were made while learning, as reminders

# to RUN APP IN CMD PROMPT: cd to this directory, or place in default CMD directory:
# then run 'python rubicon_reminders_cli.py'

from os import listdir
from datetime import datetime

# this assigns dt variable as date + timestamp
dt = (datetime.now())
# TODO numerate note items per entry


# open existing or create a new file prompt:
p1 = input("(V)iew or (N)ew [V/N]: ").upper()
if p1 == "V":
    # this views file directory of existing notes if first input is (view)
    for file in listdir():
        if file.endswith(".txt"):
            print(file)
    # below opens existing file, allows multiple note lines, and dates it when finished with session.
    old_file = (input("which file would you like to open: "))
    hdl = open(old_file + ".txt", "r+")  # using r+ by default places text at beginning, overwriting.
    for line in hdl:  # as long as you first READ the file, then r+ becomes APPEND TO END.
        print(line.strip())
    of_note = input("Add Note: ")
    if of_note == "done":  # FIXME add accept on any 'done' check for upper and lowercase
        # specifies notes were reviewed if first note entry is "done"
        hdl.write(" REVIEWED: ")
        hdl.write(str(dt))
    # if first entered note is not 'done', continue asking for more notes until entry == 'done'
    else:
        hdl.write('\n')
        hdl.write(of_note)
        hdl.write('\n')
        while of_note != "done":
            of_note = input("Add more notes: ")
            while of_note != "done":
                hdl.write(of_note)
                hdl.write('\n')
        else:
            hdl.write("SESSION END: ")
            hdl.write(str(dt))
            hdl.write('\n')
        hdl.close()

# below is the block for generating and noting in a new file, if line 15 == 'N'
elif p1 == "N":
    new_file = input("new file name: ")
    hdl = open(new_file, "a")
    nf_note = input("Add Note: ")
    if nf_note == "done":
        print("finished")
    else:
        hdl.write(nf_note)
        hdl.write('\n')
        while nf_note != "done":
            nf_note = input("Add more notes: ")
            while nf_note != "done":
                hdl.write(nf_note)
                hdl.write('\n')
                break
        else:
            hdl.write("SESSION END: ")
            hdl.write(str(dt))
            hdl.write('\n')
    hdl.close()
else:
    print("Error: please enter V or N")



main.py
# tkinter GUI functions
from tkinter import Tk, Text, Button, Listbox, END, Entry, Scrollbar, Menu, Label

# os imports for creating note directory, adding and removing text files, and changing working dir to note dir
from os import getcwd, path, listdir, chdir, remove, makedirs

# datetime used to display save function has run at time clicked
from datetime import datetime

# project module holding menu bar functions
import menu_contents

# generates note directory
import todo_tab


def directory_generator():
    current_directory = getcwd()
    final_directory = path.join(current_directory, r"note_folder")
    if not path.exists(final_directory):
        makedirs(final_directory)


directory_generator()

# changes working directory to generated note directory
chdir("note_folder")

# creates a global working_note variable to store active note, so that listbox.curselection does not break when
# double-clicking away from listbox and then trying to save file
working_note = ()


# main GUI and functions
def main_gui():
    window = Tk()
    window.geometry("400x400")
    window.config(bg="#000000")
    window.title("Rubicon Reminders")

    # function to refresh window
    def go_home():
        window.destroy()
        main_gui()

    # main listbox interface
    lbox = Listbox(window)
    lbox.place(x=10, y=40)

    # populates listbox with note file names
    for file in listdir():
        if file.endswith('.txt'):
            lbox.insert(END, file)

    # entry box for adding a new note
    add_note_entry = Entry(window)
    add_note_entry.place(x=15, y=268)
    add_note_entry.config(width=17)
    add_note_label = Label(
        window, text="Enter New Note Title:", bg="#000000", fg="#FFFFFF"
    )
    add_note_label.place(x=12, y=245)

    # function to add a named note to the listbox population
    def add_note(event):
        title = add_note_entry.get()
        hdl = open(title + ".txt", "w")
        hdl.close(), go_home()

    # function for opening and reading selected note
    def open_note(event):
        lbox_selection = lbox.curselection()[0]
        file = lbox.get(lbox_selection)
        global working_note
        working_note = file
        with open(file, "r+") as file:
            file = file.read()
        text_output.delete("1.0", END)
        text_output.insert(END, file)

    # function for writing additional text to opened file
    def save_note_cmd(event):
        file = working_note
        hdl = open(file, "w")
        note = text_output.get("1.0", "end")
        for line in note:
            hdl.write(line)
        hdl.close()
        # displays note-save functionality is working correctly
        dt = datetime.now()
        save_confirm.delete("1.0", "end")
        save_confirm.insert("1.0", "Note saved: " + str(dt))

    # function for deleting a selected note from listbox and directory
    def del_note(event):
        lbox_selection = lbox.curselection()[0]
        file = lbox.get(lbox_selection)
        remove(file)
        go_home()

    # GUI buttons
    # Buttons which share functions bound with keybind events must be called with lambda
    def buttons():
        add_note_button = Button(window, command=lambda: add_note(add_note_entry))
        add_note_button.place(x=15, y=290)
        add_note_button.config(width=14, height=1, text="ADD NEW NOTE", relief="raised")

        del_note_button = Button(window, command=lambda: del_note(lbox))
        del_note_button.place(x=8, y=210)
        del_note_button.config(
            width=17, height=1, text="DELETE SELECTED NOTE", relief="raised"
        )

        save_note_button = Button(window, command=lambda: save_note_cmd(text_output))
        save_note_button.pack(side="bottom", fill="x")
        save_note_button.config(
            height=1,
            text="SAVE NOTE",
            bg="#8E001C",
            font=("RobotoRoman", 11, "bold"),
            relief="groove",
            fg="#FFB302",
        )

        todo_button = Button(text="OPEN TODO", command=todo_tab.run)
        todo_button.config(width=18, bg="#FFB302", height=1, font=("RobotRoman", 11, "bold"))
        todo_button.pack(side="top", fill="x")

    buttons()


    # text area to display confirm message when save is successful.
    save_confirm = Text(window, height=1, width=50)
    save_confirm.pack(side="bottom")

    # text area to display selected note from listbox
    text_output = Text(window, wrap="word")
    text_output.pack(
        side="right",
        fill="y",
        pady=4
    )
    text_output.config(width=29)

    # scrollbar, tied to text_output
    text_scroll = Scrollbar(window, command=text_output.yview)
    text_scroll.pack(side="right", fill="y", pady=4)
    text_output["yscrollcommand"] = text_scroll.set

    # binds selected listbox item to be opened on selection, into text_output
    lbox.bind("<<ListboxSelect>>", open_note)

    # binds ctrl-s as hotkey to save text_output back to file
    text_output.bind("<Control-s>", save_note_cmd)

    # binds Enter key to add note once note title is typed in entry area.
    add_note_entry.bind("<Return>", add_note)

    # binds 'del' key to delete text file in lbox (selection area)
    lbox.bind('<Delete>', del_note)

    # menu bar
    menu_bar = Menu(window)
    menu_bar.add_command(label="Help", command=menu_contents.help_screen)
    menu_bar.add_command(
        label="View Note Folder", command=menu_contents.view_note_folder
    )
    window.config(menu=menu_bar)



    window.resizable(False, False)
    window.mainloop()


main_gui()
